### PR TITLE
Implement forwarding protocol

### PR DESCRIPTION
**Pull request by tsipinakis**
## Please describe the change you are making

Implement a protocol for connection forwarding

## Are you the owner of the code you are sending in, or do you have permission of the owner?

Sent with permission from the owner

## The code will be published under the MIT-0 license. Have you read and understood this license?

Yes

### CODE REVIEW

File: forwardServer.go [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+		syscall.SIGINT,
+		syscall.SIGTERM,
+		syscall.SIGQUIT)
+	go func() {
+		for {
+			_ = <-sigc
+			// ... do something ...
+			buf := make([]byte, 1<<20)
+			stacklen := runtime.Stack(buf, true)
+			str := fmt.Sprintf("=== received SIGQUIT ===\n*** goroutine dump...\n%s\n*** end\n", buf[:stacklen])
```
Comment (janosdebugs) [inline forwardServer.go:201]: We should think about a good way to serialize these in the log library. Logging this kind of stuff may prove tricky with syslog if we ever decide to add it.

File: forwardServer.go [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+	case proto.CONNECTION_TYPE_SOCKET_FORWARD:
+		fallthrough
+	case proto.CONNECTION_TYPE_PORT_FORWARD:
+		localForward(log, &forwardCtx, connChan, setup)
+	case proto.CONNECTION_TYPE_SOCKET_DIAL:
+		fallthrough
+	case proto.CONNECTION_TYPE_PORT_DIAL:
+		externalDial(log, &forwardCtx, connChan, setup)
+	}
+	//ctx := proto.ForwardCtx{
```
Comment (janosdebugs) [inline forwardServer.go:229]: !!! commented out code!

File: go.mod [hunk 3]
```diff
@@ -1,3 +1,8 @@
 module github.com/containerssh/agent
 
-go 1.16
\ No newline at end of file
+go 1.16
+
+require (
+	github.com/containerssh/libcontainerssh v0.0.0-20220313181044-57a5eba20399
+	github.com/fxamacker/cbor/v2 v2.3.0
```
Comment (janosdebugs) [inline go.mod:7]: Please use the latest version, they are working hard on stability improvements.

**Review by janosdebugs — CHANGES_REQUESTED** [review]
Looks good, but a few things may need some work.

File: protocol/Protocol.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
+package protocol
```
[the remark this answers is not in the document]
Comment (tsipinakis) [reply protocol/Protocol.go:1]: There already is a unified library for the protocol, `server.go` :)

File: forwardServer.go [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+	setup.BindHost = "127.0.0.1"
+	// Magic X11 formula: 6000 + display number
+	setup.BindPort = 6010
+	return setup
+}
+
+func parsePort(proto string, host string, port uint32) string {
+	switch proto {
+	case "tcp":
+		return host + ":" + strconv.Itoa(int(port))
```
[the remark this answers is not in the document]
Comment (tsipinakis) [reply forwardServer.go:64]: It should, host can also be a dns hostname so if there's an ipv6 address in there it should be properly escaped.

File: protocol/Protocol.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (janosdebugs) [reply protocol/Protocol.go:1]: We still need to move it to avoid a circular dependency.

**Comment by janosdebugs** [comment]
@tsipinakis now that libcontainerssh is merged can we get this sorted?

**Comment by tsipinakis** [comment]
Need to merge https://github.com/ContainerSSH/libcontainerssh/pull/142 before finishing this.

**Pull request closed without merging**